atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/transaction/impl/expectations/TrackingClosableIteratorTest.java

```diff
@@ -17,85 +17,100 @@
+        List<String> strings = ImmutableList.of("", "length1", "length2", "test");
+        ClosableIterator<String> trackingIterator =
+                createTrackingIterator(ClosableIterators.wrapWithEmptyClose(strings.iterator()));
+        assertThat(trackingIterator).toIterable().containsExactlyElementsOf(strings);
     }
 
     @Test
-    public void trackingClosableStringIteratorDelegatesClose() {
-        ClosableIterator<String> iterator = spy(createClosableStringIterator());
-        TrackingClosableIterator<String> trackingIterator =
-                new TrackingClosableIterator<>(iterator, NO_OP, STRING_LENGTH_MEASURER);
-        trackingIterator.close();
-        verify(iterator).close();
-        verifyNoMoreInteractions(iterator);
+    public void trackingIteratorFeedsTracker() {
+        when(measurer.applyAsLong(anyString())).thenReturn(1L).thenReturn(2L).thenReturn(3L);
+        ClosableIterator<String> trackingIterator =
+                createTrackingIterator(ClosableIterators.wrapWithEmptyClose(Iterators.forArray("one", "two", "three")));
+        trackingIterator.forEachRemaining(_unused -> {});
+
+        InOrder inOrder = inOrder(tracker);
+        inOrder.verify(tracker).record(1L);
+        inOrder.verify(tracker).record(2L);
+        inOrder.verify(tracker).record(3L);
+        verifyNoMoreInteractions(tracker);
     }
 
-    private static ClosableIterator<String> createClosableStringIterator() {
-        return ClosableIterators.wrapWithEmptyClose(createStringIterator());
+    @Test
+    public void trackingIteratorForwardsValuesDespiteExceptionAtMeasurement() {
+        when(measurer.applyAsLong(anyString())).thenThrow(RuntimeException.class);
+        ClosableIterator<String> trackingIterator =
+                createTrackingIterator(ClosableIterators.wrapWithEmptyClose(Iterators.singletonIterator(STRING)));
+        assertThat(trackingIterator).toIterable().containsExactly(STRING);
+    }
+
+    @Test
+    public void trackingIteratorForwardsValuesDespiteExceptionAtConsumption() {
+        doThrow(RuntimeException.class).when(tracker).record(anyLong());
+        ClosableIterator<String> trackingIterator =
+                createTrackingIterator(ClosableIterators.wrapWithEmptyClose(Iterators.singletonIterator(STRING)));
+        assertThat(trackingIterator).toIterable().containsExactly(STRING);
+    }
+
+    @Test
+    @SuppressWarnings("unchecked")
+    public void trackingClosableStringIteratorDelegatesClose() {
+        ClosableIterator<String> delegate = mock(ClosableIterator.class);
+        ClosableIterator<String> trackingIterator = createTrackingIterator(delegate);
+        trackingIterator.close();
+        verify(delegate).close();
+        verifyNoMoreInteractions(delegate);
     }
 
-    private static Iterator<String> createStringIterator() {
-        return STRINGS.stream().iterator();
+    public ClosableIterator<String> createTrackingIterator(ClosableIterator<String> delegate) {
```

Comment:
nit: every single test wraps an iterator with `ClosableIterators.wrapWithEmptyClose`. Consider having this method take in an Iterator<String> and wrap it within this method.

Reply:
Yep! :)